### PR TITLE
RegExp Comparison

### PR DESCRIPTION
**Pull request by amuniz**
Add a way to compare two regular expressions on their matching spaces.

Thanks for this awesome library @zhztheplayer!

### CODE REVIEW

File: src/main/java/top/yatt/dfargx/RegexComparator.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import top.yatt.dfargx.automata.DFA;
+import top.yatt.dfargx.automata.NFA;
+import top.yatt.dfargx.tree.SyntaxTree;
+
+public class RegexComparator {
+
+    /**
+     * Checks if the matching space of <code>regexp2</code> is contained by the matching space of <code>regexp1</code>.
+     */
+    public static final boolean isContained(String regexp1, String regexp2) {
```
Comment (zhztheplayer) [inline src/main/java/top/yatt/dfargx/RegexComparator.java]: Better to call "contains" or find regexp1 from regexp2.

File: src/main/java/top/yatt/dfargx/automata/DFA.java [hunk 2]
```diff
@@ -40,6 +40,33 @@ public int getInitState() {
         return fs;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof DFA)) {
+            return false;
+        }
+        DFA other = (DFA) o;
+        if (other.is == this.is && other.rs == this.rs) {
```
Comment (zhztheplayer) [inline src/main/java/top/yatt/dfargx/automata/DFA.java]: I guess this is based on the assumption that initial state and rejected state are always the same after compilation of "different but equal" regex patterns.

Is this assumption always working?

File: src/main/java/top/yatt/dfargx/automata/DFA.java [hunk 3]
```diff
@@ -40,6 +40,33 @@ public int getInitState() {
...
+                return false;
+            }
+            for (int i = 0; i < other.fs.length; i++) {
+                if (other.fs[i] != this.fs[i]) {
+                    return false;
+                }
+            }
+            for (int i = 0; i < other.transitionTable.length; i++) {
+                for (int j = 0; j < other.transitionTable[i].length; j++) {
+                    if (other.transitionTable[i][j] != this.transitionTable[i][j]) {
```
Comment (zhztheplayer) [inline src/main/java/top/yatt/dfargx/automata/DFA.java]: The same.

File: src/main/java/top/yatt/dfargx/RegexComparator.java (hunk 1 above)
Comment (amuniz) [reply src/main/java/top/yatt/dfargx/RegexComparator.java]: Do you mean the method name? Sure, I can change it to whatever you think is better.

File: src/main/java/top/yatt/dfargx/automata/DFA.java (hunk 2 above)
Comment (amuniz) [reply src/main/java/top/yatt/dfargx/automata/DFA.java]: Correct. If I understood correctly, a regular expression contains another if and only if their DFA's are _identical_, which means all their attributes are (including initial state and rejected state).

I'm sure your knowledge about DFA is far more deep than mine, so please correct me if I'm wrong!

File: src/main/java/top/yatt/dfargx/RegexComparator.java (hunk 1 above)
Comment (zhztheplayer) [reply src/main/java/top/yatt/dfargx/RegexComparator.java]: Yes, I mean the method name. 

IMO, "isContained" always implies that left is contained by right, and "contains" indicates the opposite meaning.

e.g
[org.apache.commons.lang3.StringUtils.contains](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html)
[org.apache.commons.collections4.CollectionUtils.containsAll](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html)

So I think either changing the method name or just swapping the parameter order are better. What do you think?

File: src/main/java/top/yatt/dfargx/automata/DFA.java (hunk 2 above)
Comment (zhztheplayer) [reply src/main/java/top/yatt/dfargx/automata/DFA.java]: Yes, the equal regex pattern should have identical DFA, but the identical DFAs can have different state numbers.

e.g.
### DFA1

|            |a   |b|
|-----------|----|---|
|1(initial) | 2   |3|
|2(accepted) |3   |3|
|3(rejected) |3   |3|

### DFA2
|            |a   |b|
|-----------|----|---|
|1(rejected) | 1  |1|
|2(initial) |3   |1|
|3(accepted) |1  |1|

These two DFAs are exactly the same and minimal, but still have state numbers different.

Maybe we can make the "equals" method cover this case?

Comment (amuniz) [reply src/main/java/top/yatt/dfargx/automata/DFA.java]: Ok, I get what you mean, and I wonder why tests are just working fine...
I could try to write a better equals method, but unfortunately I don't have the required knowledge.

Do you have a good link to some documentation example on DFA equivalence?

**Comment by amuniz** [comment]
I did some research :smile: 

I think the current approach is really comparing transition graphs properly, what do you think @zhztheplayer?

**Comment by zhztheplayer** [comment]
@amuniz 

The code looks good to me, Thanks for the contribution about this new feature!

**Comment by amuniz** [comment]
@zhztheplayer could you generate a release with this feature please?

**Comment by zhztheplayer** [comment]
@amuniz 
Still not ready for a 0.2.2 release. but you can anyway use the 0.2.2-SNAPSHOT instead, it does contain this feature.

**Pull request merged**